### DISCUSSION TITLE
Content Security Policy setting

### DISCUSSION THREAD

**Discussion by itsgifnotjiff** [community/community#49832] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Question

### Body

I am trying to ensure that an embeded Google maps `iframe` displays properly when I build and deploy to GitHub Pages and I have no idea where to set the `unsafe-inline` property for the Content Security Policy so that I do not get a `404` once deployed. Any ideas where and how I would do this in this [Vue 3/Vite static SPA](https://github.com/logue/vite-vuetify-ts-starter)? 

```typescript
<v-row>
  <div style="width: 100%">
    <iframe src="data.mapUrl" width="100%" height="450" frameborder="0" style="border:0;" loading="lazy"
      referrerpolicy="no-referrer-when-downgrade"></iframe>
  </div>
</v-row>
```

**Comment by emilburganov** [comment]
https://stackoverflow.com/questions/40866219/how-to-resolve-iframe-cross-domain-issue

**Comment by m-amin-b** [comment]
Hey! You're hitting a **classic GitHub Pages + CSP + Google Maps iframe** gotcha — and you're not alone. The `404` you're seeing after deploy is almost certainly because **GitHub Pages enforces a strict default Content Security Policy (CSP)** that blocks `unsafe-inline` styles/scripts *and* often rejects data URLs or external iframes unless explicitly allowed.

But good news: **you can fix this in your Vue 3 + Vite project with zero server config** — just add a `_headers` file (Netlify-style headers work on GitHub Pages too!).

---

### Solution: Add a `_headers` file to allow Google Maps iframe

1. **Create a file** at the root of your `public` folder:  
   `public/_headers`

2. **Add this content** (adjust if you have other rules):
   ```headers
   /*
     Content-Security-Policy: default-src 'self'; script-src 'self' 'unsafe-inline' https://*.googleapis.com https://*.gstatic.com; style-src 'self' 'unsafe-inline' https://*.google.com; frame-src https://www.google.com; img-src 'self' data: https:; connect-src 'self' https://*.googleapis.com;

**Comment by bdnath702** [comment]
You’re running into the typical GitHub Pages + CSP problem where the Google Maps iframe gets blocked after deployment. GitHub Pages uses a restrictive default CSP, which prevents inline scripts/styles and external map iframes unless you manually allow them — that’s why it works locally but fails once uploaded.
A simple workaround is to add a _headers file inside your project’s public directory. GitHub Pages understands these header rules and will apply your custom CSP.

Create this file:
public/_headers

### Add something like:
/*
Content-Security-Policy: default-src 'self';
script-src 'self' 'unsafe-inline' [](https://.googleapis.com https://.gstatic.com);
style-src 'self' 'unsafe-inline';
frame-src https://www.google.com/ https://maps.google.com/;
img-src 'self' data: https://.googleapis.com https://.ggpht.com;
This opens up just enough permissions for the Google Maps iframe to load properly.
Also make sure your iframe uses Vue binding: use :src="data.mapUrl" instead of src="data.mapUrl".

**Comment by iyawnnn** [comment]
To solve this issue and allow the embedded Google Maps iframe to display correctly when deploying your Vue 3/Vite app to GitHub Pages, you'll need to adjust the Content Security Policy (CSP) to allow loading content from Google Maps.

Here’s a straightforward approach:

Update the index.html file: In a Vue 3/Vite project, the index.html file is where the main HTML template resides. You need to add the Content-Security-Policy meta tag to this file to specify the allowed sources.

Modify the Content-Security-Policy (CSP): You need to include Google Maps in the list of trusted sources by allowing unsafe-inline and https://maps.googleapis.com. Update the CSP in your index.html like this:

```
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta http-equiv="Content-Security-Policy" content="default-src 'self'; script-src 'self' 'unsafe-inline' https://maps.googleapis.com; style-src 'self' 'unsafe-inline'; frame-src https://www.google.com;">
  <title>Your App</title>
</head>
```


Here’s what each directive does:

default-src 'self': Allows content only from your domain.

script-src 'self' 'unsafe-inline' https://maps.googleapis.com: Permits loading scripts from Google Maps and inline scripts.

style-src 'self' 'unsafe-inline': Allows inline styles.

frame-src https://www.google.com: Allows iframes from Google.

Deploy to GitHub Pages: Once you've updated the CSP, rebuild your project and deploy it to GitHub Pages. This should resolve the 404 issue and allow the Google Maps iframe to load properly.

**Comment by SayedSadiq45** [comment]
You cannot set or override a Content-Security-Policy header on GitHub Pages, and that’s why your Google Maps iframe fails after deployment. GitHub Pages enforces its own CSP and does not allow you to add unsafe-inline or custom CSP headers in a Vue/Vite static site.

Key facts

GitHub Pages does not let you configure response headers (CSP, HSTS, etc.).

You cannot add CSP rules through Vite, Vue config, or index.html → Pages ignores them.

If your embed requires unsafe-inline, GitHub Pages will block it.

Why your embed breaks

Your Google Maps iframe is using:

src="data.mapUrl"


But in Vue, that is interpreted as a string literal, not a bound value.
When built, the browser sees:

src="data.mapUrl"


—which is invalid → GitHub Pages returns 404 for that URL.

This has nothing to do with CSP.

✅ Correct way to bind Google Maps iframe in Vue 3
<iframe
  :src="mapUrl"
  width="100%"
  height="450"
  frameborder="0"
  style="border:0;"
  loading="lazy"
  referrerpolicy="no-referrer-when-downgrade"
></iframe>

❗ No CSP changes needed

Google Maps iframes work fine under GitHub Pages’ CSP as long as the src is valid.

Summary

You cannot set unsafe-inline on GitHub Pages.

Your issue is caused by incorrect Vue binding (src="data.mapUrl").

Fix the binding → iframe loads normally → no CSP header needed.

If you need, I can show you how to store the map URL, or provide a complete Vue component.
